Aggrandize Moderation - (Freeze command & other improvements)

applenick opened the pull request:
# Aggrandize Moderation
### *(or simply more improvements to moderation commands)*
I'm back once again with several great changes to the moderation feature set. Most notably re-implementing the `/freeeze` command! Also added are a few important improvements to bans to make the feature set even better. 

Resolves #135

## Freeze :snowflake:
With this reintroduced feature, moderators will now have the ability to freeze via command or observer tool. Also added is a command to view a list of frozen players.

Command: `/freeze [target]` - Freezes a player in place
Command: `/unfreeze [target]` - Unfreezes a player
Command: `/freezelist` - Displays a list of frozen players

Note: At the moment the Freeze tool does not contain any configuration values. I felt that since all of the moderation commands at this time are enabled by default, there would be no reason to make this addition any different.

A few additional details regarding this feature, I've added it so when a player logs off after being frozen, logging back on they will be re-frozen by the console. Also frozen players will be unable to perform commands while frozen except for `/msg` and `/r`. Though at this time I chose not to block frozen players from speaking in chat completely. However I would like some feedback on whether this is something that should be changed. 

## Moderation Improvements
I've included a brief overview of the more noteworthy changes to moderation features:
* Using `/ipban [username]` of a player who was recently banned, will ban their IP as we cache recently banned players.
* Fixed unbanned players triggering similar IP account alert.
* Will alert sender when an IP ban is successful (includes # of online alts banned).
* Now banned/temp-banned players will see a fully formatted kick screen when attempting to rejoin.

## Misc Changes.
* Clean up admin chat broadcasts with new static method.
* Reformat TNT defuser display name to match other tools.

## Screenshots: 
![mod-pov](https://user-images.githubusercontent.com/3377659/77995182-26f43200-72e0-11ea-8336-6516a68b4cd2.gif)
Example of using freeze tool and alert it broadcasts in the admin chat.

![frozen-pov](https://user-images.githubusercontent.com/3377659/77995287-5f940b80-72e0-11ea-8883-53cce4bc309a.png)
What the title looks like that is displayed to frozen players.

![re-log](https://user-images.githubusercontent.com/3377659/77996057-c534c780-72e1-11ea-853b-eb34916e95ca.gif)
If a frozen player re-logs, they will be frozen by the console until the match cycles or unfrozen by a staff member.

![tools-view](https://user-images.githubusercontent.com/3377659/77995308-67ec4680-72e0-11ea-8531-1533375c2872.gif)
The more consistent display name for TNT Defuser along with new Freeze tool.

![freeze-list](https://user-images.githubusercontent.com/3377659/77995317-6b7fcd80-72e0-11ea-929a-f1950c14cb32.png)
Example of what is shown to staff using the `/freezelist` command.


## Final thoughts
That should be everything that was changed in this PR. I did perform extensive testing and everything should be ready to go. Let me know if there are any suggestions/changes that need to be done and I’ll take care of it as soon as I can.

Signed-off-by: applenick <applenick@users.noreply.github.com>

Pablete1234 commented on core/src/main/java/tc/oc/pgm/community/modules/FreezeMatchModule.java:
```diff
@@ -0,0 +1,349 @@
+    ItemStack stack = new ItemStack(TOOL_MATERIAL);
+    ItemMeta meta = stack.getItemMeta();
+    meta.setDisplayName(
+        ChatColor.DARK_AQUA
```
```suggestion
        ChatColor.AQUA
```

applenick replied:
Hey @Pablete1234,
I want to keep all the tool names as separate colors, the observer tool is already Aqua. Do you have any recommendations? Perhaps bolded white or would you rather I swap the colors between the observer tools and this?

applenick replied:
I’m marking this as resolved, please check my most recent comment below for the new format. I think it looks better than dark aqua. Let me know if you agree.

KingOfSquares commented on core/src/main/java/tc/oc/pgm/community/modules/FreezeMatchModule.java:
```diff
@@ -0,0 +1,349 @@
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBucketFill(final PlayerBucketFillEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBucketEmpty(final PlayerBucketEmptyEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
```
I believe you can just cancel PlayerBucketEvent instead of both fill and empty?

applenick replied:
Good idea! I’ll test that out and let you know.

applenick replied:
Ok it’s really weird but the `PlayerBucketEvent` caused a lot of headaches for an hour. For some reason it caused an issue where players were unable to join the server at all. So I’m just going to keep the separate events.

KingOfSquares replied:
Well, thanks for trying 👍

KingOfSquares commented on core/src/main/java/tc/oc/pgm/community/modules/FreezeMatchModule.java:
```diff
@@ -0,0 +1,349 @@
+        event.setCancelled(true);
+      }
+    }
+  }
```
Same for this, why not just cancel all `InventoryEvent`s, that way to can also cancel events like `InventoryOpenEvent` and opening other interfaces like Anvils, Crafting tables..

applenick replied:
Good thinking, however I think it’s fine to allow the inventory open event. Let’s say a moderator is questioning the contents of the frozen player’s inventory, it may be useful to allow the user to open up their inventory and see. Plus with the inventory click blocking it, they can’t edit it anyway.

applenick replied:
@KingOfSquares Also just checked, InventoryEvent is not cancellable either.

KingOfSquares commented on core/src/main/java/tc/oc/pgm/community/modules/FreezeMatchModule.java:
```diff
@@ -0,0 +1,349 @@
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onVehicleDamage(final VehicleDamageEvent event) {
```
This should maybe be placed together with the other VeichleEvents? And why not just cancel all `VehicleEvent`s? :p

applenick replied:
You can’t cancel VehicleEvent xD

KingOfSquares commented on core/src/main/java/tc/oc/pgm/community/modules/FreezeMatchModule.java:
```diff
@@ -0,0 +1,349 @@
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
```
Repeating my point, but. Why not cancel all `BlockEvent`s?

EudaldC commented:
Players should be unfrozen when they leave. Sometimes they leave before they can be punished, and if they rejoin they would be frozen with nobody to unfreeze them. It was intended design in PGM 1.9+.

applenick commented on core/src/main/java/tc/oc/pgm/community/modules/FreezeMatchModule.java:
```diff
@@ -0,0 +1,349 @@
+  }
+
+  @EventHandler
+  public void giveKit(final ObserverKitApplyEvent event) {
```
@KingOfSquares I understand your idea, however that would break convention. Any tool that has built in functionality (picker, observer tools) is created in the match module where the related functionality is located. Were this featured handled outside of PGM then yeah it would be located there. Thanks for the suggestion though 😄

applenick commented:
@EudaldC thank you for the suggestion. I went ahead and modified it so now if a frozen player quits it only caches them for 3 minutes. I feel this should resolve the issue you brought up. 

Regarding what else I updated:
Changed the display name of the player freezer to a white to differentiate it. I like it more than the Dark Aqua, but still feel that Aqua fits better under the observer tool. @Pablete1234 

Also some minor adjustments such as adding offline players to `/freezelist`. Check my commit message for all the changes.

Let me know if there is anything else that can be improved!
